generator.go

```diff
@@ -96,17 +81,32 @@ func (g *DataGenerator) Generate(params *GenerateParams) (interface{}, error) {
 		requestPathDisplay = "(empty request path)"
 	}
 
-	return g.generateInternal(&GenerateParams{
+	data, err := g.generateInternal(&GenerateParams{
 		Expansions:  params.Expansions,
-		ID:          params.ID,
+		PathParams:  nil,
 		RequestPath: params.RequestPath,
 		Schema:      params.Schema,
 
-		context:     fmt.Sprintf("Responding to %s:\n", requestPathDisplay),
-		doReplaceID: true,
-		example:     nil,
-		replacedID:  nil,
+		context: fmt.Sprintf("Responding to %s:\n", requestPathDisplay),
+		example: nil,
 	})
+	if err != nil {
+		return nil, err
+	}
+
+	if params.PathParams != nil {
+		// Passses through the generated data and replaces IDs that existed in
+		// the fixtures with IDs that were extracted from the request path, if
+		// and where appropriate.
+		replaceIDs(params.PathParams, data)
+
+		// Passes through the generated data again to replace the values of any old
+		// IDs that we replaced. This is a separate step because IDs could have
+		// been found and replace at any point in the generation process.
+		distributeReplacedIDs(params.PathParams, data)
```

Comment:
This two-step structure is counterintuitive; I didn't realize that `replaceIDs` was modifying `PathParams`. A couple ideas: Could we give `replaceIDs` a more descriptive name? Could we make it return a separate object instead of modifying `PathParams`, to make the data flow more obvious? Could these comments be clearer?

Reply:
Yes, that's a great point. I ended up doing three things:

1. Renaming: `replaceIDs` is now called `recordAndReplaceIDs` to hint that it's also storing the IDs it replaces.
2. Changing the signatures. `params.PathParams` is still mutated in place, but the same struct is now returned as well just to make it clearer that information is flowing out of the first function and being used by the second.
3. Tightening up the comments.

Here's what the code looks like now:

``` ruby
	if params.PathParams != nil {
		// Passses through the generated data and replaces IDs that existed in
		// the fixtures with IDs that were extracted from the request path, if
		// and where appropriate.
		//
		// Note that the path params are mutated by the function, but we return
		// them anyway to make the control flow here more clear.
		pathParams := recordAndReplaceIDs(params.PathParams, data)

		// Passes through the generated data again to replace the values of any old
		// IDs that we replaced. This is a separate step because IDs could have
		// been found and replace at any point in the generation process.
		distributeReplacedIDs(pathParams, data)
	}
```

I think it'll help clarity a lot. Lemme know if you think of anything else I could improve, and I'll make those changes as well.